segfault

Hi 

somewhere between 44e36038f032eff2f36522642a8cc4351e133f83 ( good ) and 764c1bd5ea856dd7db5171783e442321dc3e041f (bad)  segfault when opening sql file.

unfortunately I can not post sql file here ( prod code )

I am not familiar enough with tree-sitter, whats the easiest way to run parser though any sql file ?

I've tried adding offending sql file to `test/highlight` but that did't do anything
```
syntax highlighting:
  ✓ union.sql (13 assertions)
  ✓ query.sql (14 assertions)
  ✓ structure.sql (0 assertions) <--- this is the file that causes segfault in neovim

```

I have little to no experience with C, but I have run `scanner.c` through static code analysis (cppcheck) and it reports the following:

```
src/scanner.c:29:3: error: Found an exit path from function with non-void return type that has missing return statement [missingReturn]
  }
  ^
src/scanner.c:63:5: error: Memory leak: text_size [memleak]
    return NULL;
    ^
Active checkers: 40/565
```

The second error is in the function `scan_dollar_string_tag()`. @DerekStride @dmfay any ideas how to fix that?

--- a/src/scanner.c
+++ b/src/scanner.c
@@ -28,20 +28,20 @@ void *tree_sitter_sql_external_scanner_destroy(void *payload) {
     state->start_tag = NULL;
   }
   free(payload);
+  return NULL;
 }
 
-char* add_char(char* text, int* text_size, char c, int index) {
+char* add_char(char* text, size_t* text_size, char c, int index) {
   if (text == NULL) {
     text = malloc(sizeof(char) * MALLOC_STRING_SIZE);
     *text_size = MALLOC_STRING_SIZE;
-    text[index] = c;
-    text[index + 1] = '\0';
   }
 
-  if (index + 1 == *text_size) {
+  // will break when indexes advances more than MALLOC_STRING_SIZE
+  if (index + 1 >= *text_size) {
     *text_size += MALLOC_STRING_SIZE;
     char* tmp = malloc(*text_size * sizeof(char));
-    strcpy(tmp, text);
+    strncpy(tmp, text, *text_size);
     free(text);
     text = tmp;
   }
@@ -54,12 +54,13 @@ char* add_char(char* text, int* text_size, char c, int index) {
 char* scan_dollar_string_tag(TSLexer *lexer) {
   char* tag = NULL;
   int index = 0;
-  int* text_size = malloc(sizeof(int));
+  size_t* text_size = malloc(sizeof(size_t));
   *text_size = 0;
   if (lexer->lookahead == '$') {
     tag = add_char(tag, text_size, '$', index);
     lexer->advance(lexer, false);
   } else {
+    free(text_size);
     return NULL;
   }
 
@@ -181,7 +182,7 @@ void tree_sitter_sql_external_scanner_deserialize(void *payload, const char *buf
   state->start_tag = NULL;
   // A length of 1 can't exists.
   if (length > 1) {
-    state->start_tag = malloc(length - 1);
+    state->start_tag = malloc(length);
     strcpy(state->start_tag, buffer);
   }
 }

